pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,267 @@
+var _ framework.QueueSortPlugin = &Coscheduling{}
+var _ framework.PermitPlugin = &Coscheduling{}
+var _ framework.UnreservePlugin = &Coscheduling{}
+
+// Name is the name of the plugin used in Registry and configurations.
+const Name = "coscheduling"
+const PodGroupName = "pod-group.scheduling.sigs.k8s.io/name"
+const PodGroupMinAvailable = "pod-group.scheduling.sigs.k8s.io/min-available"
+const PermitWaitingTime = time.Second
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (cs *Coscheduling) Name() string {
+	return Name
+}
+
+// New initializes a new plugin and returns it.
+func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin, error) {
+	podLister := handle.SharedInformerFactory().Core().V1().Pods().Lister()
+	podGroupInfos := make(map[string]PodGroupInfo)
+	return &Coscheduling{FrameworkHandle: handle,
+		PodLister:     podLister,
+		PodGroupInfos: podGroupInfos,
+	}, nil
+}
+
+// Less are used to sort pods in the scheduling queue.
+// 1.compare the priority of pod
+// 2.compare the timestamp of the initialization time of PodGroup
+// 3.compare the ID
+func (cs *Coscheduling) Less(podInfo1 *framework.PodInfo, podInfo2 *framework.PodInfo) bool {
+	pod1 := podInfo1.Pod
+	pod2 := podInfo2.Pod
+	priority1 := pod.GetPodPriority(pod1)
+	priority2 := pod.GetPodPriority(pod2)
+	if priority1 != priority2 {
+		return priority1 > priority2
+	}
+
+	var pgInfo1 PodGroupInfo
+	var pgInfo2 PodGroupInfo
+	var ok bool
+	podGroupName1, min1, namespace1 := cs.checkPod(pod1)
+	podGroupName2, min2, namespace2 := cs.checkPod(pod2)
+	if podGroupName1 != "" {
+		key1 := namespace1 + "/" + podGroupName1
+		pgInfo1, ok = cs.PodGroupInfos[key1]
+		if !ok {
+			pgInfo1 = PodGroupInfo{
+				InitialTimestamp: pod1.CreationTimestamp.Time,
+				Id:               cs.GetId(),
+				MinAvailable:     min1,
+			}
+			cs.PodGroupInfos[key1] = pgInfo1
+		}
+	}
+	if podGroupName2 != "" {
+		key2 := namespace2 + "/" + podGroupName2
+		pgInfo2, ok = cs.PodGroupInfos[key2]
+		if !ok {
+			pgInfo2 = PodGroupInfo{
+				InitialTimestamp: pod2.CreationTimestamp.Time,
+				Id:               cs.GetId(),
+				MinAvailable:     min2,
+			}
+			cs.PodGroupInfos[key2] = pgInfo2
+		}
+	}
+
+	if podGroupName1 != "" && podGroupName2 != "" {
+		if pgInfo1.InitialTimestamp.Equal(pgInfo2.InitialTimestamp) {
+			return pgInfo1.Id < pgInfo2.Id
+		} else {
+			return pgInfo1.InitialTimestamp.Before(pgInfo2.InitialTimestamp)
+		}
+	} else if podGroupName1 != "" {
+		return pgInfo1.InitialTimestamp.Before(podInfo2.InitialAttemptTimestamp)
+	} else if podGroupName2 != "" {
+		return podInfo1.InitialAttemptTimestamp.Before(pgInfo2.InitialTimestamp)
+	}
+
+	return podInfo1.InitialAttemptTimestamp.Before(podInfo2.InitialAttemptTimestamp)
+}
+
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+	podGroupName, minAvailable, namespace := cs.checkPod(p)
+	if podGroupName == "" || minAvailable == 0 {
```

Comment:
```suggestion
	if podGroupName == "" || minAvailable <= 1 {
```

could we shortcut when `minAvailable== 0 or 1` ??

Reply:
Cool~ I will shortcut when minAvailable== 1, thanks~